torch/csrc/jit/script/compiler.cpp

```diff
@@ -478,15 +539,63 @@ struct to_ir {
     }
   }
 
-  // This will _always_ compute something, unlike 'getValue' which simply
-  // returns an already computed reference if possible.
+  // special rules apply when we directly call foo(a,b) when foo is an ident
+  std::vector<Value*> emitApplyIdent(Ident ident, std::vector<Value*> inputs, List<Attribute> attributes, size_t output_size) {
+    auto it = function_table.find(ident.name());
+    if (it != function_table.end()) {
+      if(inputs.size() != it->second.num_inputs())
+        throw ErrorReport(ident) << "expected " << it->second.num_inputs() << " but found " << inputs.size();
+      auto outputs = method.emit_call_to(it->second, inputs);
+      expectOutputs(ident, output_size, outputs.size());
+      return outputs;
+    } else if (ident.name() == "print") {
+      expectOutputs(ident, output_size, 0);
+      if (!attributes.empty())
+        throw ErrorReport(ident) << "print doesn't accept any keyword arguments";
+      return emitNode(kPrint, ident.range(), inputs, 0 )->outputs();
+    }
+    Node* builtin = emitBuiltinCall(ident.range(), method, ident.name(), inputs, attributes, output_size);
+    if (hasTensorOp(builtin)) {
+      return builtin->outputs();
+    }
+    builtin->destroy();
+    // it wasn't known built in, so treat it like standard apply
+    return emitApplyExpr(Var::create(ident.range(), ident), inputs, attributes, output_size);
+  }
+
+  std::vector<Value*> emitApplyExpr(Expr callee, const std::vector<Value*>& inputs, List<Attribute> attributes, size_t output_size) {
+    // otherwise we evaluate the callee and then desugar it
+    auto sv = emitSugaredExpr(callee);
+    return sv->call(callee.range(), method, inputs, attributes, output_size);
+  }
+
+  // any expression that can produce a SugaredValue are handled here
+  // with emitExpr falling back to this function to handle them
+  // the kinds handled here should be kept in sync with [SUGARED VALUES]
+  // in emitExpr
+  std::shared_ptr<SugaredValue> emitSugaredExpr(Expr tree) {
+    switch(tree.kind()) {
+      case TK_VAR:
+        return environment_stack->getSugaredVar(Var(tree).name());
+      case '.': {
+        auto select = Select(tree);
+        auto sv = emitSugaredExpr(select.value());
+        return sv->attr(select.range(), method, select.selector().name());
+      }
+      default:
+        return std::make_shared<SimpleValue>(emitExpr(tree, 1)[0]);
```

Comment:
`(a + 1).sum()` will cause the call `emitSugaredExpr` on `a + 1`. The infinite recursion does not happen because the only time `emitExpr` calls `emitSugaredExpr` is when it knows there is a sugared Expr. I will try refactoring into 3 functions so it is easier to see that there is no recursive cycle, and to make it more likely it stays that way.

Reply:
I know the recursion is impossible right now, but it’s quite easy to break it in the future. I just meant that it can be refactored in a safer way